example/client/cmd/util.go

```diff
@@ -50,23 +54,120 @@ func deleteAPIClient(cs *clientState) {
 	cs.connection = nil
 }
 
-func parseAttributes(attributes string) (map[string]string, error) {
-	attrs := make(map[string]string)
-	if len(attributes) > 0 {
-		for _, a := range strings.Split(attributes, ",") {
+type assignFunc func(index int32, value string) error
+
+func decomposeAttributes(attrs string, dict map[int32]string, assign assignFunc) error {
+	if len(attrs) > 0 {
+		for _, a := range strings.Split(attrs, ",") {
 			i := strings.Index(a, "=")
 			if i < 0 {
-				return nil, fmt.Errorf("Attribute value %v does not include an = sign", a)
+				return fmt.Errorf("Attribute value %v does not include an = sign", a)
 			} else if i == 0 {
-				return nil, fmt.Errorf("Attribute value %v does not contain a valid name", a)
+				return fmt.Errorf("Attribute value %v does not contain a valid name", a)
 			}
 			name := a[0:i]
 			value := a[i+1:]
-			attrs[name] = value
+			index := int32(len(dict))
+			dict[index] = name
+
+			err := assign(index, value)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+func parseAttributes(rootArgs *rootArgs) (*mixerpb.Attributes, error) {
+	attrs := mixerpb.Attributes{}
+	attrs.Dictionary = make(map[int32]string)
+
+	// once again, the following boilerplate would be more succinct with generics...
+
+	if err := decomposeAttributes(rootArgs.stringAttributes, attrs.Dictionary, func(index int32, value string) error {
```

Comment:
whoa. can we move the func() up a line or something (maybe pull it out entirely) to make this bit easier to follow?

(same question for the rest of these).